Core functionality for fx variables handling in volume and area statistics in light of PR 439 (PR 439 child 2)

valeriupredoi opened the pull request:
Before you start, please read [CONTRIBUTING.md](https://github.com/ESMValGroup/ESMValCore/blob/master/CONTRIBUTING.md).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [ ] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [ ] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [ ] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

**READ BEFORE PROCEEDING**

Issue #436 lays down the need for preprocessing fx variables as integrated with `area_atatistics` and `volume_statistics`; PR #439 implements that; reviewers expressed their consternation that that PR is huuuge so it got broken down; the layout for the PR 439 children is outlined in https://github.com/ESMValGroup/ESMValCore/issues/436#issuecomment-589717672
This is the 2nd child which contains the bulk of the new functionality.

**Features**

- allows for fx variables preprocessing tasks to be ancestors of the main variable task;
- runs the same preprocessing steps in the main variable preprocessor on each specified fx variable, asked for area and volume statistics (*note*: bar the time preprocessors)
- writes the Preprocessor files for the fx vars in the preproc directory of the main variable
- allows for data loading from that dir, as if the fx preprocessed file was attached to the main variable from the get going
- implements a filter on ancestors with identical names (within one task) to eliminate duplication

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -352,71 +352,86 @@ def _get_default_settings(variable, config_user, derive=False):
+    # add special ensemble for CMIP5 only
     if fx_variable['project'] == 'CMIP5':
-        fx_variable['mip'] = 'fx'
         fx_variable['ensemble'] = 'r0i0p0'
```
yes, now `mip`s are automatically retrived at line 386 (`# Get all fx-related mips ('fx' always first, original mip last)`) - this is something @schlunma has innovated so we can have automation for cases when mip should be eg Oyr instead of Ofx - a very useful concoction :beer:

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -352,71 +352,86 @@ def _get_default_settings(variable, config_user, derive=False):
 
+    # force only the mip declared by user
+    if 'mip' in fx_variable:
+        fx_mips = [fx_variable['mip']]
```
@schlunma care to tell Klaus the logic behind? I can say that this is a safeguard to account for all possible permutations

schlunma replied:
> 
> 
> I also wonder if we can take the non-fx tables into account here: If the variable does not live in, eg, `Ofx`, but `Omon` then that will also be the `mip` of `variable`.

I'm not 100% sure if I got you correctly. Are you proposing to just search in every table instead of only the one related to `variable`? Right now, this only searches in the non-fx table with the `mip` of `variable`.

valeriupredoi replied:
...and that's about as safe as one can go at the moment, w/o going through all the tables, as Manuel points out

zklaus replied:
There are two things here: One ist the code suggestion and the other is my comment. Let's deal with the suggestion first. Do you agree that it is functionally equivalent?

valeriupredoi replied:
OK plugged the code change :+1:

zklaus replied:
Right, on to the comment. Maybe I misunderstand the code, but here is how I read it:

If the user has explicitly provided a `mip`(table) for the "fx" variable, use that. Otherwise, try to find the requested "fx" variable in all tables of the project (CMIP5 or CMIP6) that contain the literal string "fx", starting with `fx` itself.
If neither happens, presumably there will be an error down the line.

Now my suggestion stems from the observation that the most reasonable (maybe the only) other candidate for the table is the table of the data variable itself. For example, `volcello` for `(Omon, thetao)` can either come from `Omon` or `Ofx`. `Odec` would be unreasonable.
Hence, why not start the list of guessing tables with the table of the data variable?

Does that make sense?

schlunma replied:
Yes, I agree, haven't thought about this thoroughly.

So I guess we can start the `else` loop with

```python
fx_mips = [variable['mip'], 'fx']
```

and drop the last line in the loop.

zklaus replied:
Ok, I'll add a new suggestion to that effect.

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -352,71 +352,86 @@ def _get_default_settings(variable, config_user, derive=False):
+        valid_fx_vars = valid_fx_vars[0]
 
-    return fx_files
+    return fx_files, valid_fx_vars
```
I know...this will probably go away soon when iris implements full functionality for cell measures anyway...

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -845,6 +899,16 @@ def append(group_prefix, var):
     return derive_input
 
 
+def _remove_time_preproc(fxprofile):
```
It's a bit confusing that this function doesn't modify in place. Maybe a name like `_get_filtered_fxprofile` would be better?

valeriupredoi replied:
done :beer:

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -845,6 +899,16 @@ def append(group_prefix, var):
+    return fxprofile
+
+
 def _get_preprocessor_task(variables, profiles, config_user, task_name):
```
90 lines and 7 levels of indentation in one function? Please refactor.

valeriupredoi replied:
K boss, refactored it by taking the fx ancestry up into a separate function so we can easily remove when we get rid of ancestry -> @bouweandela

valeriupredoi replied:
done :beer:

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -893,14 +1002,32 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
+            for anc_product in ancestor_task.products:
+                ancestors_filename_dict[anc_product.filename] = ancestor_task
+                filenames.append(anc_product.filename)
+        set_ancestors = list(
```
Why the conversion to list?

valeriupredoi replied:
removed it, good call

valeriupredoi replied:
done :beer:

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -981,10 +1108,10 @@ def _initialize_datasets(raw_datasets):
+        """Expand ensemble members to multiple datasets.
+
+        Expansion only support ensembles defined as strings, not lists.
 
```
Why the empty line?

valeriupredoi replied:
pep8, not me, doing yapf -i

valeriupredoi replied:
done :beer:

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -352,71 +352,86 @@ def _get_default_settings(variable, config_user, derive=False):
+        fx_mips = ['fx']
+        fx_mips.extend(
+            [key for key in cmor_table.tables if 'fx' in key and key != 'fx'])
+        fx_mips.append(variable['mip'])
```
```suggestion
        fx_mips = [variable['mip'], 'fx']
        fx_mips.extend(
            [key for key in cmor_table.tables if 'fx' in key and key != 'fx'])
```
As discussed.

valeriupredoi replied:
done :beer:

valeriupredoi commented:
review totally appreciated @zklaus - I'll address the suggestions today :beer:

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -495,15 +495,17 @@ def _update_fx_settings(settings, variable, config_user):
                         raise RecipeError(
-                            "Key short_name missing from {}".format(fxvar))
+                            "No valid short_name found in {}".format(fxvar)
+                        )
```
Sorry for not having been clear enough. I presented two alternatives, now we have a mixture that is again a bit unclear since actually with the try-except an invalid but present entry could slip by.

So to clarify: Either catch `KeyError` -> "No entry found" OR test for validity -> "No valid name found".

valeriupredoi replied:
bit confused, how does it look now?

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -475,13 +492,27 @@ def _update_fx_settings(settings, variable, config_user):
+
+            # finally construct the fx files dictionary
             settings[step]['fx_files'] = fx_files_dict
             logger.info(msg, step, pformat(fx_files_dict))
```
there is no actual duplication, if you look carefully, there are a lot of differences between the cases both at logger and functional levels. However, I broke down the big function in two parts now, that deal with totally separate things

zklaus replied:
There absolutely **is** duplication: line 474 is the same as 491, line 476 is almost the same as 492 and it smells strongly like they should be identical. Indeed, the structure I proposed above would go some way towards that since the `step_settings` would deal with the differences in the lvalue. That leaves the difference that one assigns a dict, the other a list of dicts. Furthermore, the code of `_get_landsea_fraction_fx_dict` looks suspiciously similar to the code block in the `'mask_landseaice'` branch. Probably all three cases can be dealt with in one function similar to an extended version of `_get_landsea_fraction_fx_dict` that takes a list of fx variables as an additional argument.

valeriupredoi replied:
gonna try change now :beer:

valeriupredoi replied:
what do you think about it now @zklaus -man?

zklaus replied:
I put a more fleshed out suggestion in #522. Have a look and tell me what you think. :beers:

valeriupredoi replied:
Shall do in a wee bit, need coffee now. Cheers, man :beer:

valeriupredoi replied:
#522 merged in here so am resolving the convo :beer:

zklaus replied:
Ok. I'm putting the open conversation in the comments here then.

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -665,13 +696,21 @@ def get_matching(attributes):
 
 def _get_preprocessor_products(variables, profile, order, ancestor_products,
-                               config_user):
+                               config_user, fx_case=False):
```
it is a flag that determines the coupling between the main`var` or `variable` and the fx variable - a kludge (or hack) that's needed to let the workflow know that it needs info from both and pass it to the data finder and juggle the task creating correctly. This will disappear once we remove the ancestry for fx vars preprocessing (I renamed it to `var_fxvar_coupling` for now to make it more relevant)

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -740,13 +784,23 @@ def _get_single_preprocessor_task(variables,
+            ancestor_products=ancestor_products,
+            config_user=config_user,
+            fx_case=True
+        )
```
yep, total rookie mistake here :grin:

valeriupredoi replied:
done :beer:

valeriupredoi commented on tests/integration/test_recipe.py:
```diff
@@ -1891,10 +1891,366 @@ def test_fx_vars_volcello_in_ofx_cmip6(tmp_path, patched_datafinder,
     assert len(fx_files) == 1
-    assert '_Ofx_' in fx_files['volcello']
+    assert '_Omon_' in fx_files['volcello']
+    assert '_Ofx_' not in fx_files['volcello']
```
@schlunma note this change due to https://github.com/ESMValGroup/ESMValCore/pull/511#discussion_r383785460 - is this right?

valeriupredoi replied:
so you happy with the result @schlunma ? :beer:

valeriupredoi replied:
cheerios, dudios :beer:

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -352,71 +352,86 @@ def _get_default_settings(variable, config_user, derive=False):
     """Get fx files (searching all possible mips)."""
-    # TODO: allow user to specify certain mip if desired
+    # make it a dict
+    if not isinstance(fx_variable, dict):
```
done!

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -475,13 +492,27 @@ def _update_fx_settings(settings, variable, config_user):
+        fx_files_dict = {}
         if settings.get(step, {}).get('fx_files'):
-            var = dict(variable)
             var['fx_files'] = settings.get(step, {}).get('fx_files')
```
done :beer:

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -665,13 +696,21 @@ def get_matching(attributes):
+        for variable in variables:
+            variable['filename'] = get_output_file(parent_variable,
+                                                   config_user['preproc_dir'],
+                                                   fx_var_alias=variable)
```
nope, that needs to stay a list of variables for some reason, otherwise the `filename` key does not get assigned

zklaus replied:
To me, this looks equivalent. Could you please point out my mistake?

valeriupredoi replied:
because I am redefining the whole set of `variables` in this case, to be used below for updating all the settings and building the products:
```
    if not var_fxvar_coupling:
        for variable in variables:
            variable['filename'] = get_output_file(variable,
                                                   config_user['preproc_dir'])
    else:
        parent_variable = variables[1]
        variables = [variables[0]]
        for variable in variables:
            variable['filename'] = get_output_file(parent_variable,
                                                   config_user['preproc_dir'],
                                                   fx_var_alias=variable)

    if ancestor_products:
        grouped_ancestors = _match_products(ancestor_products, variables)
    else:
        grouped_ancestors = {}

    for variable in variables:
        settings = _get_default_settings(
            variable,
            config_user,
            derive='derive' in profile,
        )
...
```
variables are now `[variables[0]]`

valeriupredoi replied:
no mistake on yer part K-man, just a reassignment of variables->variables, you ok with it?

zklaus replied:
Now I get it. I'm not super happy with the loop over clearly only one element (and am also wondering if there ever could be more than just two elements in the original variables), but :man_shrugging:.

valeriupredoi replied:
ya, looping on one element is peasant-y but it's the path of least changes I could find; the `variables` in the non-fx case could be a list of multiple elements, but not for fx since the task is built per fx variable. This will go to when we eliminate the ancestors, but until then it's the best I can muster for correct functionality :beer:

zklaus replied:
I'm ok(ish) with the code as it stands. I made one last suggestion by updating the suggestion above. If you like, take it, otherwise leave as is. :beers:

valeriupredoi replied:
dude, you a hero - cheers for all the work, done and put this one last change :beer:

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -694,9 +733,9 @@ def _get_preprocessor_products(variables, profile, order, ancestor_products,
-            config_user=config_user)
+        _update_fx_settings(settings=settings,
+                            variable=variable,
+                            config_user=config_user)
```
done :beer:

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -665,13 +696,21 @@ def get_matching(attributes):
 
 
 def _get_preprocessor_products(variables, profile, order, ancestor_products,
```
done :beer:

valeriupredoi commented:
OK so due to all these changes, even if the tests cover most of the thing, I retested with a comprehensive functional test (a big recipe) and all is well :+1:

valeriupredoi commented on esmvalcore/_data_finder.py:
```diff
@@ -249,12 +249,20 @@ def get_output_file(variable, preproc_dir):
+            variable['diagnostic'],
+            variable['variable_group'],
+            _replace_tags(cfg['output_file'], fx_var_alias)[0],
+        )
```
this is done, so am resolving the convo

valeriupredoi commented:
@zklaus @bouweandela dudes what's your say - we are starting to rely even more on this to be merged and am aware of its shortcomings and kluges but at the moment it'll prove more useful if merged than sitting in a corner - can we try and see about this? Muchos gracias :beer:

zklaus commented:
That really underscores the importance of smaller and cleaner PRs to begin with. After all, nobody needs a code review to know that 90 lines of code and 7 levels of indentation in one function are not appropriate.

Anyway, I'll copy here the open discussion from the side PR #522:

@zklaus wrote:
> Next step: unify `_update_fx_settings_mask` and `_update_fx_settings_stats`.
> 
> For this it might be helpful to note that `_get_spatial_stats_fx_vars` really has nothing about spatial or stats things, but rather just gets the requested fx vars from the setting.
> 
> Looking at line 534 we notice that when the update method is called, the settings parameter already contains the pertinent subsection so that the manual distinction in lines 520, 524, and 528 is unnecessary.
> 
> Indeed, this automatic determination of fx variables makes one wonder if that could replace the manual listing of `sft{lf,of,gif}` entirely.

@valeriupredoi wrote:
> but `_get_spatial_stats_fx_vars` needs to be called because it's the only function that gets the `fx_files` attr from the step (the others - mask landsea and weighted - don't need to). Kinda hard to unify the two funcs coz one need the call to `get_output_file` - ideas? beer

@zklaus wrote:
> Well, in one we place the fx file as determined by `_get_correct_fx_file` in the `fx_dict`, in the other the result of `get_output_file`. What's the reason for this difference?

@valeriupredoi wrote:
> `get_output_file` is used as a patch to get the already preprocessed and written to disk fx file, since the preprocessing and file writing is done independently of the whole fx workflow and is done like for any other variable

@zklaus wrote:
> But why is that needed for some files, but not for others? What if a custom order is applied and both mask and area files have been regridded, for instance?

valeriupredoi commented:
> But why is that needed for some files, but not for others? What if a custom order is applied and both mask and area files have been regridded, for instance?

That is a very good question! The workflow for fx files that are used for area/volume/zonal_statistics is fundamentally different and herein lies the tricksy nature of this PR (well, the part of it that deals with preprocessing these data):

- the input fx data needed for those preprocessors are preprocessed entirely and firstly (as ancestors to the main variable) and these preprocessed files are written to disk, the very first thing the workflow does;
- then, and only then, after all work's been done on those files only do the area/volume/zonal_statistics preproc step should look for files for fx_files - and those are U-turned from the get_output_files() assuming (and knowing) that they are there

This is a kludge - just because variable-fxvariable coupling is unavailable in a true sense now, given the current architecture. @bouweandela and myself we can work next using iris' new cell_measure functionality so we can do this in an elegant and practical way. Note that this method is relatively foolproof ie it is not prone to fail unexpectedly - if it fails, it fails due to an issue in the preprocessing step that's easy to trace.

About regridding - at the moment both the mask and model data will be regridded which is bad, but then again we can always put the area/volume/zonal_statistics preprocessor before the regridding step; we can also include the regrid as a prohibited step for the fx data just as are the time preproc steps. I left that out so I can talk to you about it - and here we are, talking about it :grin:

valeriupredoi commented:
also to note: the preprocessor steps applied to fx data are a subset of the main variable they're associated with: [step1, step2...stepN] where stepN is the last step just before area/volume/zonal_statistics, everything that follows after that for the main var preprocessor is not applied to the fx variable; also step1-N **do not** include any time preprocessors

zklaus commented:
Really like these last changes at a glance! Will do in-depth review tomorrow. Of course, the `for`-loop needs to be turned into a dict comprehension.

valeriupredoi commented:
> Really like these last changes at a glance! Will do in-depth review tomorrow. Of course, the `for`-loop needs to be turned into a dict comprehension.

cheers dude, done it, and also simplified and added test - now the user can define his own fx variables no matter what mask preproc

valeriupredoi commented:
this is a terribly outdated branch, together with the other sisters, they awere useful to @ledm and @LisaBock I believe, but I think it's about time to retire them (close them) - what do you think @zklaus and @bouweandela - do we maybe want to recycle some of this functionality towards implementing the cell measures stuffs with new iris?

bouweandela commented:
Maybe keep it around until #999 is merged?

schlunma commented:
I believe we can close this since #999 got merged, can't we?

zklaus commented:
Since V wanted to close them for a long time, I'll do it now. @valeriupredoi, @bouweandela please reopen if you disagree.
